### DISCUSSION TITLE
After installing via Homebrew, system fails to use the newer version

### DISCUSSION THREAD

**Discussion by ghost** [Homebrew/discussions#373] [Getting started]
Hi,

I'm on Big Sur and have installed Homebrew.

My $PATH contains:
```
% tr ':' '\n' <<< "$PATH"
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
/Applications/VMware Fusion.app/Contents/Public
/Library/TeX/texbin
```

But for example, after installing git, I get the following
```
% which git
/usr/bin/git
```
and
```
% git --version
git version 2.24.3 (Apple Git-128)
```

Is my $PATH correct?

**Comment by gromgit** [comment]
Did you `brew install git`? If not, there's no Homebrew git to show.

`brew ls` will show you what you've already installed via Homebrew.

**Reply by ghost** [reply]
Sure, `brew install git` is what I did.

**Reply by gromgit** [reply]
And what does `brew ls` output?

Also, `brew info git` will give you more detailed info about the git you installed.

**Reply by ghost** [reply]
```
aom			glib			libffi			nettle			sdl2
bdw-gc			gmp			libidn2			nghttp2			snappy
c-ares			gnutls			libogg			opencore-amr		speex
cairo			gobject-introspection	libpng			openjpeg		sqlite
dav1d			graphite2		libsamplerate		openssl@1.1		srt
ffmpeg			guile			libsndfile		opus			tesseract
flac			harfbuzz		libsoxr			p11-kit			theora
fontconfig		icu4c			libtasn1		pcre			unbound
freetype		jemalloc		libtiff			pcre2			webp
frei0r			jpeg			libtool			pixman			x264
fribidi			lame			libunistring		pkg-config		x265
gdbm			leptonica		libvidstab		python@3.9		xvid
gettext			libass			libvorbis		rav1e			xz
ghostscript		libbluray		libvpx			readline		youtube-dl
giflib			libev			little-cms2		rtmpdump
git			libevent		lzo			rubberband
anki                          iina                          notion                        visual-studio-code
firefox                       iterm2                        rectangle                     vmware-fusion
google-backup-and-sync        mactex-no-gui                 spyder
```
I have the same behavior with python
```
% python --version
Python 2.7.16
```

**Reply by gromgit** [reply]
> I have the same behavior with python

`brew info python` and read the Caveats section.

**Reply by gromgit** [reply]
Since `brew ls` lists `git`, something else is amiss.

What do `brew config` and `brew doctor` output?

**Reply by ghost** [reply]
Python has been installed as
  /usr/local/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /usr/local/opt/python@3.9/libexec/bin

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.9/site-packages

**Reply by ghost** [reply]
brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Homebrew's sbin was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting the PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc

**Reply by ghost** [reply]
I did `export PATH="/usr/local/opt/python/libexec/bin:$PATH"` before but it disappears

What do I need to do after installing Homebrew for things like python, pip etc to work?

**Reply by gromgit** [reply]
> Warning: Homebrew's sbin was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.

Assuming you installed Homebrew via the official installer script:
```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
it would've printed a **Next Steps** at the end that had instructions on what else to do. One of those steps would've looked like this:
```
==> Next steps:
[...]
- Add Homebrew to your PATH in /home/aho/.profile:
    echo 'eval $(/usr/local/bin/brew shellenv)' >> /home/aho/.profile
    eval $(/usr/local/bin/brew shellenv)
[...]
```
Use `brew shellenv` to set up all the necessary Homebrew environment variables, and you'll never have to wonder if you set your `PATH` correctly.

> I did `export PATH="/usr/local/opt/python/libexec/bin:$PATH"` before but it disappears

Typing that in a shell session only changes your `PATH` for that shell session. You need to add it to your shell startup file (I'm guessing `~/.zshrc`) if you want it to be permanent.

> What do I need to do after installing Homebrew for things like python, pip etc to work?

The Python stuff is documented in the Python caveat above. For everything else...we still haven't seen your `brew config` and `brew info git` output.

**Reply by ghost** [reply]
Thanks for the clarification mate. Until now, I used Homebrew mostly installing, latex, imagemagick, youtube-dl or casks
Thus, this problems never occurred before.

I tried `echo 'eval $(/usr/local/bin/brew shellenv)' >> /home/aho/.profile` but I got `zsh: no such file or directory: /home/aho/.profile`
And I can't remember the exact output after the installation script of Homebrew
Do I need to do the above commands also every time I start a new instance or write it into `~/.zshrc` ?
Because my $PATH reads now:
```
/usr/local/sbin
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
/Applications/VMware Fusion.app/Contents/Public
/Library/TeX/texbin
```
And I **still** get for `python --version` this: `Python 2.7.16`

Here is `brew config`
```
HOMEBREW_VERSION: 2.7.0
ORIGIN: https://github.com/Homebrew/brew
HEAD: 993bf95877143afe1cede022083499e0b4fe336c
Last commit: 6 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 00ca6afc5e94edae057956856c48f92be22aaac9
Core tap last commit: 2 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 12.0 build 1200
Git: 2.29.2 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.1-x86_64
CLT: 12.3.0.0.1.1607026830
Xcode: N/A
```
Here is `brew info git`
```
git: stable 2.29.2 (bottled), HEAD
Distributed revision control system
https://git-scm.com
/usr/local/Cellar/git/2.29.2 (1,480 files, 39.7MB) *
  Poured from bottle on 2020-12-27 at 08:31:33
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/git.rb
License: GPL-2.0-only
==> Dependencies
Required: gettext ✔, pcre2 ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
The Tcl/Tk GUIs (e.g. gitk, git-gui) are now in the `git-gui` formula.

zsh completions and functions have been installed to:
  /usr/local/share/zsh/site-functions

Emacs Lisp files have been installed to:
  /usr/local/share/emacs/site-lisp/git
==> Analytics
install: 130,625 (30 days), 564,024 (90 days), 2,291,169 (365 days)
install-on-request: 128,792 (30 days), 546,937 (90 days), 2,195,296 (365 days)
build-error: 0 (30 days)
```

**Reply by gromgit** [reply]
> I tried `echo 'eval $(/usr/local/bin/brew shellenv)' >> /home/aho/.profile` but I got `zsh: no such file or directory: /home/aho/.profile`

Since you're using Zsh and don't have `~/.profile`, I guess `~/.zshrc` is the most appropriate place.

> And I still get for `python --version` this: Python 2.7.16

As the Python caveat you posted says:
```
Python has been installed as
/usr/local/bin/python3

Unversioned symlinks python, python-config, pip etc. pointing to
python3, python3-config, pip3 etc., respectively, have been installed into
/usr/local/opt/python@3.9/libexec/bin
```
_Versioned_ Homebrew Python 3 (i.e. `python3`) is in `/usr/local/bin`, which is in the standard Homebrew `PATH`.

_Unversioned_ Homebrew Python 3 (i.e. `python`) is in `/usr/local/opt/python@3.9/libexec/bin`, which is _not_ in the standard Homebrew `PATH` to avoid confusion with unversioned system Python (2).

This is why you keep getting the same result for:
```
% which python
/usr/bin/python
```

If you specifically need Python 3 for your scripts, it's _much_ safer to run the versioned `python3`, `pip3`, etc. than rely on unversioned `python`, `pip`, etc.

**Comment by ghost** [comment]
To be more specific:
```
==> Caveats
Python has been installed as
  /usr/local/bin/python3
```
I doesn't make sense to me that the system is using `/usr/bin/` rather than `/usr/local/bin/` which is above in $PATH (pls look at the comment above)
```
% which python
/usr/bin/python
```

**Comment by Linus-XZX** [comment]
Are you possibly on Apple Silicon? Homebrew installs into `/opt/homebrew` instead of `/usr/local/Homebrew` when it finds ARM, so it might not be `/usr/local` (I'm on Intel, so I'm not really sure), and Rosetta 2 could probably complicate this (judging from your python3 brew installation).
